fast_wkw/src/wkw_dataset.rs

```diff
@@ -0,0 +1,104 @@
+use lru::LruCache;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex, RwLock};
+use wkwrap::{Dataset, Header, Result, Vec3};
+
+use crate::wkw_file::File;
+
+#[derive(Clone, Debug)]
+pub struct FileCache {
+  inner: Arc<Mutex<LruCache<PathBuf, Arc<RwLock<File>>>>>,
+}
+
+impl FileCache {
+  pub fn new(cap: usize) -> Self {
+    Self {
+      inner: Arc::new(Mutex::new(LruCache::new(cap))),
+    }
+  }
+
+  pub fn get_file(&self, path: &Path) -> Option<Arc<RwLock<File>>> {
+    let mut cache = self.inner.lock().unwrap();
+    let cached_file = cache.get(&PathBuf::from(path)).cloned();
+    if let Some(cached_file) = cached_file {
+      return Some(cached_file);
+    }
+    match File::open(&path) {
+      Ok(file) => {
+        let file = Arc::new(RwLock::new(file));
+        cache.put(PathBuf::from(path), file.clone());
+        Some(file)
+      }
+      Err(_) => None,
+    }
+  }
+
+  pub fn clear_prefix(&self, path_prefix: &Path) {
+    let mut cache = self.inner.lock().unwrap();
+    let paths_to_delete = cache
+      .iter()
+      .filter_map(|(path, _)| {
+        if path.starts_with(path_prefix) {
+          Some(path)
+        } else {
+          None
+        }
+      })
+      .cloned()
+      .collect::<Vec<_>>();
+    for path in paths_to_delete.into_iter() {
+      cache.pop(&path);
+    }
+  }
+}
+
+#[derive(Clone, Debug)]
+pub struct CachedDataset {
+  root: PathBuf,
+  header: Header,
+  file_cache: FileCache,
+}
+
+impl CachedDataset {
+  pub fn new(root: &Path, file_cache: FileCache) -> Result<Self> {
+    if !root.is_dir() {
+      return Err(format!("Dataset root {:?} is not a directory", &root));
+    }
+
+    // read required header file
+    let header = Dataset::new(root)?.header().clone();
+
+    Ok(Self {
+      root: root.to_owned(),
+      header,
+      file_cache,
+    })
+  }
+
+  pub fn header(&self) -> &Header {
+    &self.header
+  }
+
+  pub fn read_block(&self, src_pos: Vec3, buf: &mut [u8]) -> Result<()> {
```

Comment:
What do you think about making this function async with `file.read().unwrap().read_block(file_src_pos, buf).await`

Reply:
I think that conflicts with the memory-mapped file approach. Reading from disk by a page fault is a blocking operation afaik. Putting everything in a thread pool should help make that non-blocking. Lz4 decoding is technically also a blocking operation, but not as problematic because it is very quick.